File: package/src/cursor.js
const combiningMarks =
    /(<%= allExceptCombiningMarks %>)(<%= combiningMarks %>+)/g,
    surrogatePair = /(<%= highSurrogates %>)(<%= lowSurrogates %>)/g;

// unicode-aware string reverse
export function reverse(str) {
    str = str.replace(combiningMarks, function (match, capture1,
        capture2) {
        return reverse(capture2) + capture1;
    })
        .replace(surrogatePair, "$2$1");
    let res = "";
    for (let i = str.length - 1; i >= 0; --i) {
        res += str[i];
    }
    return res;
}

export class Cursor {

    static min(a, b) {
        if (a.i <= b.i) {
            return a;
        }
        return b;
    }

    static max(a, b) {
        if (a.i > b.i) {
            return a;
        }
        return b;
    }

    constructor(i, x, y) {
        this.i = i || 0;
        this.x = x || 0;
        this.y = y || 0;
        this.moved = true;
        Object.seal(this);
    }

    clone() {
        return new Cursor(this.i, this.x, this.y);
    }

    toString() {
        return "[i:" + this.i + " x:" + this.x + " y:" + this.y + "]";
    }

    copy(cursor) {
        this.i = cursor.i;
        this.x = cursor.x;
        this.y = cursor.y;
        this.moved = false;
    }

    fullHome() {
        this.i = 0;
        this.x = 0;
        this.y = 0;
        this.moved = true;
    }

    fullEnd(rows) {
        this.i = 0;
        let lastLength = 0;
        for (let y = 0; y < rows.length; ++y) {
            const row = rows[y];
            lastLength = row.stringLength;
            this.i += lastLength;
        }
        this.y = rows.length - 1;
        this.x = lastLength;
        this.moved = true;
    }

    left(rows, skipAdjust = false) {
        if (this.i > 0) {
            --this.i;
            --this.x;
            if (this.x < 0) {
                --this.y;
                const row = rows[this.y];
                this.x = row.stringLength - 1;
            }
            else if (!skipAdjust) {
                rows[this.y].adjust(this, -1);
            }
        }
        this.moved = true;
    }

    skipLeft(rows) {
        if (this.x <= 1) {
            this.left(rows);
        }
        else {
            const x = this.x - 1,
                row = rows[this.y],
                word = reverse(row.substring(0, x)),
                m = word.match(/\w+/),
                dx = m
                    ? (m.index + m[0].length + 1)
                    : this.x;
            this.i -= dx;
            this.x -= dx;
            rows[this.y].adjust(this, -1);
        }
        this.moved = true;
    }

    right(rows, skipAdjust = false) {
        const row = rows[this.y];
        if (this.y < rows.length - 1
            || this.x < row.stringLength) {
            ++this.i;
            ++this.x;
            if (this.y < rows.length - 1
                && this.x === row.stringLength) {
                this.x = 0;
                ++this.y;
            }
            else if (!skipAdjust) {
                rows[this.y].adjust(this, 1);
            }
        }
    }

    skipRight(rows) {
        const row = rows[this.y];
        if (this.x < row.stringLength - 1) {
            const x = this.x + 1,
                subrow = row.substring(x),
                m = subrow.match(/\w+/),
                dx = m
                    ? (m.index + m[0].length + 1)
                    : (row.stringLength - this.x);
            this.i += dx;
            this.x += dx;
            if (this.x > 0
                && this.x === row.stringLength
                && this.y < rows.length - 1) {
                --this.x;
                --this.i;
            }
            rows[this.y].adjust(this, 1);
        }
        else if(this.y < rows.length -1) {
            this.right(rows);
        }
        this.moved = true;
    }

    home() {
        this.i -= this.x;
        this.x = 0;
        this.moved = true;
    }

    end(rows) {
        const row = rows[this.y];
        let dx = row.stringLength - this.x;
        if (this.y < rows.length - 1) {
            --dx;
        }
        this.i += dx;
        this.x += dx;
        this.moved = true;
    }

    up(rows, skipAdjust = false) {
        if (this.y > 0) {
            --this.y;
            const row = rows[this.y],
                dx = Math.min(0, row.stringLength - this.x - 1);
            this.x += dx;
            this.i -= row.stringLength - dx;
            if (!skipAdjust) {
                rows[this.y].adjust(this, 1);
            }
        }
        this.moved = true;
    }

    down(rows, skipAdjust = false) {
        if (this.y < rows.length - 1) {
            const prevRow = rows[this.y];
            ++this.y;
            this.i += prevRow.stringLength;

            const row = rows[this.y];
            if (this.x >= row.stringLength) {
                let dx = this.x - row.stringLength;
                if (this.y < rows.length - 1) {
                    ++dx;
                }
                this.i -= dx;
                this.x -= dx;
            }
            if (!skipAdjust) {
                rows[this.y].adjust(this, 1);
            }
        }
        this.moved = true;
    }

    incX(rows, dx) {
        const dir = Math.sign(dx);
        dx = Math.abs(dx);
        if (dir === -1) {
            for (let i = 0; i < dx; ++i) {
                this.left(rows, true);
            }
            rows[this.y].adjust(this, -1);
        }
        else if (dir === 1) {
            for (let i = 0; i < dx; ++i) {
                this.right(rows, true);
            }
            rows[this.y].adjust(this, 1);
        }
    }

    incY(rows, dy) {
        const dir = Math.sign(dy);
        dy = Math.abs(dy);
        if (dir === -1) {
            for (let i = 0; i < dy; ++i) {
                this.up(rows, true);
            }
        }
        else if (dir === 1) {
            for (let i = 0; i < dy; ++i) {
                this.down(rows, true);
            }
        }
        rows[this.y].adjust(this, 1);
    }

    setXY(rows, x, y) {
        x = Math.floor(x);
        y = Math.floor(y);
        this.y = Math.max(0, Math.min(rows.length - 1, y));
        const row = rows[this.y];
        this.x = Math.max(0, Math.min(row.stringLength, x));
        this.i = this.x;
        for (let i = 0; i < this.y; ++i) {
            this.i += rows[i].stringLength;
        }
        if (this.x > 0
            && this.x === row.stringLength
            && this.y < rows.length - 1) {
            --this.x;
            --this.i;
        }
        rows[this.y].adjust(this, 1);
        this.moved = true;
    }

    setI(rows, i) {
        this.x = this.i = i;
        this.y = 0;
        let total = 0,
            row = rows[this.y];
        while (this.x > row.stringLength) {
            this.x -= row.stringLength;
            total += row.stringLength;
            if (this.y >= rows.length - 1) {
                this.i = total;
                this.x = row.stringLength;
                this.moved = true;
                break;
            }
            ++this.y;
            row = rows[this.y];
        }

        if (this.y < rows.length - 1
            && this.x === row.stringLength) {
            this.x = 0;
            ++this.y;
        }

        rows[this.y].adjust(this, 1);
        this.moved = true;
    }
}

File: csharp/Primrose.NETStandard2/Cursor.cs
using System;
using System.Collections.Generic;
using System.Text.RegularExpressions;

namespace Primrose
{
    public class Cursor
    {

        private static Regex WordBoundaryPattern = new Regex("\\w+", RegexOptions.Compiled);

        public static Cursor min(Cursor a, Cursor b)
        {
            if (a.i <= b.i)
            {
                return a;
            }
            return b;
        }

        public static Cursor max(Cursor a, Cursor b)
        {
            if (a.i > b.i)
            {
                return a;
            }
            return b;
        }

        public Cursor()
            : this(0, 0, 0)
        { }

        public int i { get; set; }
        public int x { get; set; }
        public int y { get; set; }

        public Cursor(int i, int x, int y)
        {
            this.i = i;
            this.x = x;
            this.y = y;
        }

        public Cursor clone()
        {
            return new Cursor(i, x, y);
        }

        public override string ToString()
        {
            return $"[i:{i} x:{x} y:{y}]";
        }

        public void copy(Cursor cursor)
        {
            i = cursor.i;
            x = cursor.x;
            y = cursor.y;
        }

        public void fullHome()
        {
            i = 0;
            x = 0;
            y = 0;
        }

        public void fullEnd(List<Row> rows)
        {
            i = 0;
            var lastLength = 0;
            for (var y = 0; y < rows.Count; ++y)
            {
                var row = rows[y];
                lastLength = row.stringLength;
                i += lastLength;
            }
            y = rows.Count - 1;
            x = lastLength;
        }

        public void left(List<Row> rows, bool skipAdjust = false)
        {
            if (i > 0)
            {
                --i;
                --x;
                if (x < 0)
                {
                    --y;
                    var row = rows[y];
                    x = row.stringLength - 1;
                }
                else if (!skipAdjust)
                {
                    rows[y].adjust(this, -1);
                }
            }
        }

        public void skipLeft(List<Row> rows)
        {
            if (x <= 1)
            {
                left(rows);
            }
            else
            {
                var row = rows[y];
                var word = row.substring(0, x - 1).Reverse();
                var m = WordBoundaryPattern.Match(word);
                var dx = m.Success
                        ? (m.Index + m.Groups[0].Value.Length + 1)
                        : x;
                i -= dx;
                x -= dx;
                rows[y].adjust(this, -1);
            }
        }

        public void right(List<Row> rows, bool skipAdjust = false)
        {
            var row = rows[y];
            if (y < rows.Count - 1
                || x < row.stringLength)
            {
                ++i;
                ++x;
                if (y < rows.Count - 1
                    && x == row.stringLength)
                {
                    x = 0;
                    ++y;
                }
                else if (!skipAdjust)
                {
                    rows[y].adjust(this, 1);
                }
            }
        }

        public void skipRight(List<Row> rows)
        {
            var row = rows[y];
            if (x < row.stringLength - 1)
            {
                var subrow = row.substring(x + 1);
                var m = WordBoundaryPattern.Match(subrow);
                var dx = m.Success
                        ? (m.Index + m.Groups[0].Value.Length + 1)
                        : (row.stringLength - x);
                i += dx;
                x += dx;
                if (x > 0
                    && x == row.stringLength
                    && y < rows.Count - 1)
                {
                    --x;
                    --i;
                }
                rows[y].adjust(this, 1);
            }
            else if (y < rows.Count - 1)
            {
                right(rows);
            }
        }

        public void home()
        {
            i -= x;
            x = 0;
        }

        public void end(List<Row> rows)
        {
            var row = rows[y];
            var dx = row.stringLength - x;
            if (y < rows.Count - 1)
            {
                --dx;
            }
            i += dx;
            x += dx;
        }

        public void up(List<Row> rows, bool skipAdjust = false)
        {
            if (y > 0)
            {
                --y;
                var row = rows[y];
                var dx = Math.Min(0, row.stringLength - x - 1);
                x += dx;
                i -= row.stringLength - dx;
                if (!skipAdjust)
                {
                    rows[y].adjust(this, 1);
                }
            }
        }

        public void down(List<Row> rows, bool skipAdjust = false)
        {
            if (y < rows.Count - 1)
            {
                var prevRow = rows[y];
                ++y;
                i += prevRow.stringLength;

                var row = rows[y];
                if (x >= row.stringLength)
                {
                    var dx = x - row.stringLength;
                    if (y < rows.Count - 1)
                    {
                        ++dx;
                    }
                    i -= dx;
                    x -= dx;
                }
                if (!skipAdjust)
                {
                    rows[y].adjust(this, 1);
                }
            }
        }

        public void incX(List<Row> rows, int dx)
        {
            var dir = Math.Sign(dx);
            dx = Math.Abs(dx);
            if (dir == -1)
            {
                for (var i = 0; i < dx; ++i)
                {
                    left(rows, true);
                }
                rows[y].adjust(this, -1);
            }
            else if (dir == 1)
            {
                for (var i = 0; i < dx; ++i)
                {
                    right(rows, true);
                }
                rows[this.y].adjust(this, 1);
            }
        }

        public void incY(List<Row> rows, int dy)
        {
            var dir = Math.Sign(dy);
            dy = Math.Abs(dy);
            if (dir == -1)
            {
                for (var i = 0; i < dy; ++i)
                {
                    up(rows, true);
                }
            }
            else if (dir == 1)
            {
                for (var i = 0; i < dy; ++i)
                {
                    down(rows, true);
                }
            }
            rows[y].adjust(this, 1);
        }

        public void setXY(List<Row> rows, int x, int y)
        {
            this.y = y = Math.Max(0, Math.Min(rows.Count - 1, y));
            var row = rows[y];
            this.x = x = Math.Max(0, Math.Min(row.stringLength, x));
            i = x;
            for (var i = 0; i < y; ++i)
            {
                i += rows[i].stringLength;
            }
            if (x > 0
                && x == row.stringLength
                && y < rows.Count - 1)
            {
                this.x = --x;
                --i;
            }
            rows[y].adjust(this, 1);
        }

        public void setI(List<Row> rows, int i)
        {
            var delta = i - this.i;
            var dir = Math.Sign(delta);
            x = this.i = i;
            y = 0;
            var total = 0;
            var row = rows[this.y];
            while (x > row.stringLength)
            {
                x -= row.stringLength;
                total += row.stringLength;
                if (y >= rows.Count - 1)
                {
                    this.i = i = total;
                    x = row.stringLength;
                    break;
                }
                ++y;
                row = rows[y];
            }

            if (y < rows.Count - 1
                && x == row.stringLength)
            {
                x = 0;
                ++y;
            }

            rows[y].adjust(this, dir);
        }
    }
}